proto/rill/admin/v1/api.proto

```diff
@@ -2280,3 +2357,64 @@ message MemberUsergroup {
   google.protobuf.Timestamp created_on = 4;
   google.protobuf.Timestamp updated_on = 5;
 }
+
+message BillingError {
+  string organization = 1;
+  BillingErrorType type = 2;
+  BillingErrorMetadata metadata = 3;
+  google.protobuf.Timestamp event_time = 4;
+  google.protobuf.Timestamp created_on = 5;
+}
+
+message BillingErrorMetadata {
+  oneof metadata {
+    BillingErrorMetadataNoPaymentMethod no_payment_method = 1;
+    BillingErrorMetadataNoBillableAddress no_billable_address = 2;
+    BillingErrorMetadataInvoicePaymentFailed invoice_payment_failed = 3;
+    BillingErrorMetadataTrialEnded trial_ended = 4;
+    BillingErrorMetadataSubscriptionCancelled subscription_cancelled = 5;
+  }
+}
+
+message BillingErrorMetadataNoPaymentMethod {}
+
+message BillingErrorMetadataNoBillableAddress {}
+
+message BillingErrorMetadataInvoicePaymentFailed {
+  repeated InvoicePaymentFailedMeta invoices = 1;
+}
+
+message InvoicePaymentFailedMeta {
+  string invoice_id = 1;
+  string invoice_number = 2;
+  string invoice_url = 3;
+  string amount_due = 4;
+  string currency = 5;
+  google.protobuf.Timestamp due_date = 6;
+}
+
+message BillingErrorMetadataTrialEnded {
+  google.protobuf.Timestamp grace_period_end_date = 1;
+}
+
+message BillingErrorMetadataSubscriptionCancelled {
+  google.protobuf.Timestamp end_date = 1;
+}
```

Comment:
Just wondering from a simplicity point of view whether this info could be simplified to a string message? Or does the CLI/frontend need it destructured like this?

Reply:
Earlier I added string messages but found that this to be would be better for UI if they want to create call to action. For example failed invoice meta contains url that can directly take them to the failed invoice page and they can pay. Or use some of the end dates to perform something.

Reply:
For a CTA, couldn't that be managed with either just an error code and message, or message and a `cta_url`? Or does the frontend need deeper understanding of the error?

Reply:
Still feel its better to have it like this. UI can choose to create their own message. Another example on how these fields can be used is trial ending soon metadata contains trial end date or sub cancel error metadata contains sub end date so the UI can choose to display some banner or pop up on the end date or sometime before end date etc. Don't think it will be complicated for UI to use these to create flows as per product requirement since its typed metadata instead of json blobs.